### PR TITLE
Enable removal of a student from a course

### PR DESCRIPTION
**Pull request by donny-wong**
## Proposed Changes
Bulk uploading students to a course with an incorrect CSV column order results in incorrect students being created. This is one reason why this feature was implemented, which is, enabling the deletion of a single student from a course.

In order to allow the deletion of a student, they would need to not be associated to any groups, meaning there should not be any records in the `memberships` table for this student for the course they are being deleted from.

...

<details>
<summary>Screenshots of your changes (if applicable)</summary>
For testing, I created a new course with a new student:
new student creation
I then associated this student to a group:
associated student to a group
I then tried to delete the student, but got the following error:
error in deletion
I then disassociated the student from the group and was able to delete the student:
successful deletion of student

</details>

<details>
<summary>Associated <a href="https://github.com/MarkUsProject/Wiki">documentation repository</a> pull request (if applicable)</summary>
<br>
Instructions to remove a student: https://github.com/MarkUsProject/Wiki/pull/232
</details>

## Type of Change
*(Write an `X` or a brief description next to the type or types that best describe your changes.)*

| Type                                                                                    | Applies? |
|-----------------------------------------------------------------------------------------|----------|
| 🚨 *Breaking change* (fix or feature that would cause existing functionality to change) |          |
| ✨ *New feature* (non-breaking change that adds functionality)                          |     x     |
| 🐛 *Bug fix* (non-breaking change that fixes an issue)                                  |          |
| 🎨 *User interface change* (change to user interface; provide screenshots)              |          |
| ♻️ *Refactoring* (internal change to codebase, without changing functionality)          |          |
| 🚦 *Test update* (change that *only* adds or modifies tests)                            |          |
| 📦 *Dependency update* (change that updates a dependency)                               |          |
| 🔧 *Internal* (change that *only* affects developers or continuous integration)         |          |

## Checklist
*(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)*

Before opening your pull request:

- [x] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [x] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [x] I have updated the project documentation, if applicable.
  - This is **required** for new features.
- [x] If this is my first contribution, I have added myself to the list of contributors.

After opening your pull request:

- [x] I have updated the project Changelog (this is required for all changes).
- [x] I have verified that the pre-commit.ci checks have passed.
- [x] I have verified that the CI tests have passed.
- [x] I have reviewed the test coverage changes reported by Coveralls.
- [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments
*(Include any questions or comments you have regarding your changes.)*

### CODE REVIEW

File: config/locales/defaults/responders/en.yml [hunk 1]
```diff
@@ -34,6 +34,10 @@ en:
     students:
       create:
         error: "%{resource_name} could not be created due to the following error(s): %{errors}."
+      destroy:
+        error: 'Error deleting Student (%{user_name}): %{message}'
```
Comment (david-yz-liu) [inline config/locales/defaults/responders/en.yml:38]: Two comments overall:

1. Don't capitalize "Student"
2. Let's change the verb "delete" to "remove". I think that has a more accurate connotation. You can make the wording change for TAs as well.

File: spec/controllers/students_controller_spec.rb [hunk 2]
```diff
@@ -168,6 +168,52 @@
...
     end
+
+    describe '#destroy' do
+      context 'when destroy fails' do
+        before do
+          allow_any_instance_of(Role).to receive(:destroy).and_return(false)
+          delete_as instructor, :destroy, params: { course_id: course.id, id: student.id }
+        end
+
+        it 'does not delete the student, shows an error message, and gets a bad request response' do
```
Comment (david-yz-liu) [inline spec/controllers/students_controller_spec.rb:179]: these should be split up into separate tests (same below)

File: app/javascript/Components/student_table.jsx [hunk 3]
```diff
@@ -178,12 +196,19 @@ class RawStudentTable extends React.Component {
...
+                <>
+                  <span>
+                    <a href={Routes.edit_course_student_path(this.props.course_id, data.value)}>
+                      {I18n.t("edit")}
+                    </a>
+                  </span>
+                  &nbsp;|&nbsp;
+                  <span>
+                    <a href="#" onClick={() => this.removeStudent(data.value)}>
+                      {I18n.t("delete")}
```
Comment (david-yz-liu) [inline app/javascript/Components/student_table.jsx:208]: Similar to my comment below: please add a new "Remove" string and use that instead (here and with TAs)

File: app/controllers/students_controller.rb [hunk 4]
```diff
@@ -135,6 +135,22 @@ def update_settings
...
+      @role.destroy!
+    rescue ActiveRecord::DeleteRestrictionError => e
+      flash_message(:error, I18n.t('flash.students.destroy.restricted', user_name: @role.user_name, message: e.message))
+      head :conflict
+    rescue ActiveRecord::RecordNotDestroyed => e
+      flash_message(:error, I18n.t('flash.students.destroy.error', user_name: @role.user_name, message: e.message))
+      head :bad_request
+    else
+      flash_now(:success, I18n.t('flash.students.destroy.success', user_name: @role.user_name))
+      head :ok
```
Comment (david-yz-liu) [inline app/controllers/students_controller.rb:150]: `:no_content` is more appropriate here (since no content is being returned). See https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Methods/DELETE.

I'm not sure we did this for the TA destroy, but if not you can switch that one too!

File: Changelog.md [hunk 5]
```diff
@@ -12,6 +12,7 @@
 - Render download test result button on assignment summary page only if the assignment has automated testing (#7417)
 - Group test results by Test Groups id (#7422)
 - Display HTML previews of RMarkdown files (#7394)
+- Enable deletion of a student from a course (#7480)
```
Comment (david-yz-liu) [inline Changelog.md:15]: as below, please change "deletion" to "removal"

**Review by david-yz-liu — COMMENTED** [review]
@donny-wong excellent work! This PR is good, I just left a few stylistic comments.

Please also make an update to the MarkUs wiki.

**Review by david-yz-liu — COMMENTED** [review]
@donny-wong make sure to link the Wiki PR to this one (there's a section for doing so in the PR description).

Also I realized that there should be one other restriction on student deletion: if there are associated `grade_entry_student`s, and the `grade_entry_student` has an associated `grade` with a non-nil grade, then we should also prevent deletion. I expect the easiest way to implement this is to set the `grade_entry_student` association to `dependent: :destroy` and then in the grade entry student model, a check for non-nil grades.

**Comment by donny-wong** [comment]
> @donny-wong make sure to link the Wiki PR to this one (there's a section for doing so in the PR description).
> 
> Also I realized that there should be one other restriction on student deletion: if there are associated `grade_entry_student`s, and the `grade_entry_student` has an associated `grade` with a non-nil grade, then we should also prevent deletion. I expect the easiest way to implement this is to set the `grade_entry_student` association to `dependent: :destroy` and then in the grade entry student model, a check for non-nil grades.

@david-yz-liu, I am not able to delete the student `role`, regardless of having `dependent: :destroy`, as it still raises the following error on the DB level:

```
ActiveRecord::NotNullViolation - PG::NotNullViolation: ERROR:  null value in column "role_id" of relation "grade_entry_students" violates not-null constraint
2025-04-10T15:08:09.662828042Z DETAIL:  Failing row contains (450, f, 2025-04-10 15:06:43.54127, 2025-04-10 15:06:43.54127, 21, null).:
```

I would like to get your thoughts if it would be ok to remove the DB foreign key constraint of `role_id` in the `grade_entry_students` table, which would then allow me to proceed with the deletion and use `dependent: :destroy`. I would create a migration for this foreign constraint removal, and add the foreign key constraint on the `Model` level. I'm not sure if there is another way to go about this? Thank you.

**Comment by david-yz-liu** [comment]
@donny-wong you don't need to remove the foreign key constraint. The point of doing the destroy is that the associated records (i.e., grade entry forms) should be deleted as well. If this is not happening there may be something else going on that you need to investigate and fix.

**Comment by donny-wong** [comment]
> @donny-wong you don't need to remove the foreign key constraint. The point of doing the destroy is that the associated records (i.e., grade entry forms) should be deleted as well. If this is not happening there may be something else going on that you need to investigate and fix.

@david-yz-liu , you are correct. I was putting the `dependent: :destroy` in the parent `role` instead of the `student` subclass, which was overwriting it. It all makes sense now. Thank you!

**Review by david-yz-liu — COMMENTED** [review]
@donny-wong please add a test case for when there is a grade entry form (and `GradeEntryStudent`) but all of the grades are `nil`.

Also I'm not sure where in the code you are checking for non-`nil` grades before deletion?

**Comment by donny-wong** [comment]
> @donny-wong please add a test case for when there is a grade entry form (and `GradeEntryStudent`) but all of the grades are `nil`.
> 
> Also I'm not sure where in the code you are checking for non-`nil` grades before deletion?

@david-yz-liu I am checking for non-`nil` grades before deleting in `app/models/grade_entry_student.rb line 27`

**Comment by david-yz-liu** [comment]
@donny-wong ah yes I missed that (line 27), thanks for clarifying 👍

**Review by david-yz-liu — APPROVED** [review]
